## PROJECT OVERVIEW :

Spectrum Shades LLC is a prominent supplier of concrete color solutions, offering a wide range of pigments and coloring systems used in various concrete applications, including decorative concrete, precast concrete, and concrete pavers. The company prides itself on delivering high-quality colorants that meet the unique needs of its diverse clientele, including contractors, architects, and construction companies.

The company has recently observed a growing number of customer complaints regarding inconsistent color quality in their products. The discrepancies have led to a decline in customer satisfaction and a potential increase in product returns. By identifying and mitigating the factors causing color variations, the company can enhance product reliability, reduce customer complaints, and minimize return rates.

Being part of the data analysis team tasked with providing actionable insights to help Spectrum Shades LLC address the issues of inconsistent color quality and improve customer satisfaction. 

**Lets Start Our CLeaning and Transformation Journey with some Functional Programming in Python**


## Importing & Understanding The Raw Data

In [1]:
import pandas as pd
import numpy as np

data = pd.read_csv("production_data.csv")


In [2]:
data.head(5)

index  batch_id production_date  raw_material_supplier pigment_type  \
0      0         1      2024-06-25                      1       type_a   
1      1         2      2023-11-23                      2       type_b   
2      2         3      2024-02-18                      2       type_b   
3      3         4      2023-11-11                      1       type_a   
4      4         5      2024-04-11                      2       type_c   

   pigment_quantity  mixing_time mixing_speed  product_quality_score  
0         42.822881        22.56         High               7.165102  
1         42.873479        44.97         High               6.849126  
2         33.251206        52.62         High               5.661209  
3         52.532817        24.96         High               6.991735  
4         36.205108        25.37         High               7.095043

In [3]:
data.drop(columns=['index'], inplace=True, errors='ignore')
data.head(5)

batch_id production_date  raw_material_supplier pigment_type  \
0         1      2024-06-25                      1       type_a   
1         2      2023-11-23                      2       type_b   
2         3      2024-02-18                      2       type_b   
3         4      2023-11-11                      1       type_a   
4         5      2024-04-11                      2       type_c   

   pigment_quantity  mixing_time mixing_speed  product_quality_score  
0         42.822881        22.56         High               7.165102  
1         42.873479        44.97         High               6.849126  
2         33.251206        52.62         High               5.661209  
3         52.532817        24.96         High               6.991735  
4         36.205108        25.37         High               7.095043

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   batch_id               2000 non-null   int64  
 1   production_date        2000 non-null   object 
 2   raw_material_supplier  2000 non-null   int64  
 3   pigment_type           2000 non-null   object 
 4   pigment_quantity       2000 non-null   float64
 5   mixing_time            1982 non-null   float64
 6   mixing_speed           2000 non-null   object 
 7   product_quality_score  2000 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 125.1+ KB


In [5]:
data.describe()

batch_id  raw_material_supplier  pigment_quantity  mixing_time  \
count  2000.000000            2000.000000       2000.000000  1982.000000   
mean   1000.500000               1.650000         38.346191    35.278396   
std     577.494589               0.477089          6.826539    14.452281   
min       1.000000               1.000000         19.413042     3.520000   
25%     500.750000               1.000000         33.296950    23.075000   
50%    1000.500000               2.000000         37.667982    35.730000   
75%    1500.250000               2.000000         43.100851    47.525000   
max    2000.000000               2.000000         58.727253    64.180000   

       product_quality_score  
count            2000.000000  
mean                6.684824  
std                 1.387717  
min                 2.227802  
25%                 5.643263  
50%                 6.599894  
75%                 7.692736  
max                10.000000

In [6]:
def unique_values_summary(df):
    summary = {}
    for col in df.columns:
        unique_values = df[col].unique()
        summary[col] = {
            'Unique Count': df[col].nunique(),
            'Unique Values': unique_values[:10]  # Displaying only first 10 unique values 
        }
    return summary

# Get unique values summary
unique_summary = unique_values_summary(data)
print("\nUnique Values Summary:")
print(unique_summary)



Unique Values Summary:
{'batch_id': {'Unique Count': 2000, 'Unique Values': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=int64)}, 'production_date': {'Unique Count': 363, 'Unique Values': array(['2024-06-25', '2023-11-23', '2024-02-18', '2023-11-11',
       '2024-04-11', '2024-02-14', '2023-08-25', '2024-04-26',
       '2024-07-06', '2023-11-21'], dtype=object)}, 'raw_material_supplier': {'Unique Count': 2, 'Unique Values': array([1, 2], dtype=int64)}, 'pigment_type': {'Unique Count': 4, 'Unique Values': array(['type_a', 'type_b', 'type_c', 'Type_C'], dtype=object)}, 'pigment_quantity': {'Unique Count': 2000, 'Unique Values': array([42.82288066, 42.87347936, 33.25120551, 52.53281707, 36.20510832,
       29.99190302, 35.94143917, 40.49720309, 36.01511066, 42.89316877])}, 'mixing_time': {'Unique Count': 1648, 'Unique Values': array([22.56, 44.97, 52.62, 24.96, 25.37, 45.96, 32.13,  7.75,   nan,
       18.61])}, 'mixing_speed': {'Unique Count': 4, 'Unique Values': array(['High',

## Dataset Description 
The dataset *production_data.csv* contains details about production batches, raw material suppliers, pigment usage, and final product quality. However, data inconsistencies such as missing values, incorrect formats, and outliers need to be addressed.
Considering the above Data Exploration The Following Issues has been adressed.
| Column Name            | Description                  | Cleaning Criteria |
|------------------------|----------------------------|-------------------|
| **batch_id**           | Unique batch identifier    | No missing values allowed & a change in datatype is needed.|
| **production_date**    | Date of production         | Convert to datetime format |
| **raw_material_supplier** | Supplier category (1 or 2) | Map to labels: `national_supplier` / `international_supplier`, replace missing with `national_supplier` |
| **pigment_type**       | Type of pigment used       | Valid values: `type_a`, `type_b`, `type_c`, otherwise replace with `other` |
| **pigment_quantity**   | Pigment weight in kg      | Range: 1-100 kg, replace missing with median |
| **mixing_time**        | Mixing duration in minutes | Replace missing with mean |
| **mixing_speed**       | Mixing speed category      | Valid values: `Low`, `Medium`, `High`, replace missing with `Not Specified` |
| **product_quality_score** | Quality score (1-10)    | Replace missing with mean |


### Cleaning & Transformation 
 Considering the above Data Description and cleaning Criteria,We will use functional programming to clean our dataset.

In [7]:
# Function to clean the dataset
def clean_data(data):
    # Convert production_date to datetime format
    data['production_date'] = pd.to_datetime(data['production_date'], errors='coerce')
    
    # Handle missing values in raw_material_supplier
  
    data['raw_material_supplier'].fillna('national_supplier', inplace=True)
    data['raw_material_supplier'] = data['raw_material_supplier'].map({1: 'national_supplier', 2: 'international_supplier'})
    
    # Ensure pigment_type contains only valid categories, replace missing values with 'other'
    data['pigment_type'] = data['pigment_type'].str.strip().str.lower()
    data['pigment_type'].fillna('other', inplace=True)
    
    # Ensure pigment_quantity is within valid range and handle missing values
    data['pigment_quantity'] = data['pigment_quantity'].clip(1, 100)
    median_pigment_quantity = data['pigment_quantity'].median()
    data['pigment_quantity'].fillna(median_pigment_quantity, inplace=True)
    
    # Replace missing mixing_time with mean
    mean_mixing_time = data['mixing_time'].mean()
    data['mixing_time'].fillna(mean_mixing_time, inplace=True)
    
    # Handle missing values in mixing_speed
    data['mixing_speed'].replace('-', 'Not Specified', inplace=True)
    data['mixing_speed'].fillna('Not Specified', inplace=True)
    
    # Replace missing product_quality_score with mean
    data['product_quality_score'].fillna(data['product_quality_score'].mean(), inplace=True)
    
    # Ensure correct data types
    data['pigment_quantity'] = data['pigment_quantity'].astype(float).round(2)
    data['mixing_time'] = data['mixing_time'].astype(float).round(2)
    data['product_quality_score'] = data['product_quality_score'].astype(float).round(1)
    
    return data

In [8]:
Analysis_Ready_data = clean_data(data)
Analysis_Ready_data

batch_id production_date   raw_material_supplier pigment_type  \
0            1      2024-06-25       national_supplier       type_a   
1            2      2023-11-23  international_supplier       type_b   
2            3      2024-02-18  international_supplier       type_b   
3            4      2023-11-11       national_supplier       type_a   
4            5      2024-04-11  international_supplier       type_c   
...        ...             ...                     ...          ...   
1995      1996      2024-03-10  international_supplier       type_c   
1996      1997      2024-05-10  international_supplier       type_b   
1997      1998      2024-05-21  international_supplier       type_a   
1998      1999      2024-02-15  international_supplier       type_b   
1999      2000      2024-04-10       national_supplier       type_c   

      pigment_quantity  mixing_time mixing_speed  product_quality_score  
0                42.82        22.56         High                    7.2  
1                42.87        44.97         High                    6.8  
2                33.25        52.62         High                    5.7  
3                52.53        24.96         High                    7.0  
4                36.21        25.37         High                    7.1  
...                ...          ...          ...                    ...  
1995             42.92        31.59          Low                    5.0  
1996             32.26        53.31          Low                    7.6  
1997             34.33        16.45       Medium                    6.2  
1998             46.15        54.79          Low                    5.1  
1999             34.68        24.36         High                    9.5  

[2000 rows x 8 columns]

## Data Analysis

### Lets Look into the data for some Insights
##### 1. Lets see how the supplier type and quantity of materials affect the final product attributes.

In [9]:
aggregated_data = Analysis_Ready_data.groupby('raw_material_supplier').agg(
    avg_product_quality_score=('product_quality_score', 'mean'),
    avg_pigment_quantity=('pigment_quantity', 'mean')
).reset_index()

# Round the results to 2 decimal places
aggregated_data = aggregated_data.round(2)
aggregated_data


raw_material_supplier  avg_product_quality_score  avg_pigment_quantity
0  international_supplier                       5.97                 34.91
1       national_supplier                       8.02                 44.73

**INSIGHTS :**

National suppliers are associated with higher product quality scores.

National suppliers also tend to use more pigment in production.

There may be a correlation between higher pigment usage and better product quality.

##### 2.Some Data Filteration
To get more insight into the factors behind product quality,Lets Filter the data to see an average product Quality score for a specified set of results.

Identify the average product_quality_score for batches with a raw_material_supplier of 2 and a pigment_quantity greater than 35 kg.

In [11]:
file='production_data.csv'

data_2=pd.read_csv(file)

data_2new= data_2[(data_2['raw_material_supplier'] == 2)&(data_2['pigment_quantity'] > 35)]

avg_product_quality_score=data_2new['product_quality_score'].mean()

avg_pigment_quantity=data_2new['pigment_quantity'].mean()

pigment_data = pd.DataFrame({'raw_material_supplier': [2],'pigment_quantity': [round(avg_pigment_quantity, 2)],'avg_product_quality_score': [round(avg_product_quality_score, 2)]})

pigment_data.reset_index(drop=True,inplace=True)
pigment_data

raw_material_supplier  pigment_quantity  avg_product_quality_score
0                      2             39.01                       5.97

**INSIGHTS :**
International suppliers with a higher pigment quantity (>35) have an average product quality score of 5.97, which may indicate a correlation between pigment usage and quality.

The average pigment quantity is 39.01, slightly above the filtering threshold.

In [12]:
# Calculate mean and standard deviation for pigment_quantity and product_quality_score
product_quality_score_mean = data['product_quality_score'].mean()
product_quality_score_sd = data['product_quality_score'].std()
pigment_quantity_mean = data['pigment_quantity'].mean()
pigment_quantity_sd = data['pigment_quantity'].std()

# Calculate Pearson correlation coefficient between pigment_quantity and product_quality_score
corr_coef = data[['pigment_quantity', 'product_quality_score']].corr().iloc[0, 1]

# Create a DataFrame with the results
product_quality = pd.DataFrame({
    "product_quality_score_mean": [round(product_quality_score_mean, 2)],
    "product_quality_score_sd": [round(product_quality_score_sd, 2)],
    "pigment_quantity_mean": [round(pigment_quantity_mean, 2)],
    "pigment_quantity_sd": [round(pigment_quantity_sd, 2)],
    "corr_coef": [round(corr_coef, 2)]
})

# Display the results
product_quality

product_quality_score_mean  product_quality_score_sd  \
0                        6.68                      1.39   

   pigment_quantity_mean  pigment_quantity_sd  corr_coef  
0                  38.35                 6.83       0.48

**FINAL INSIGHTS :**

The product quality scores are fairly consistent (low SD).

The pigment quantity shows more variation (higher SD).

There is a moderate positive relationship (0.48 correlation) between pigment quantity and product quality, suggesting that increasing pigment quantity may improve product quality to some extent.